File: services/auth/src/domain/shared/DomainEvent.ts
// Base class for events the domain emits. Pure — no framework, no
// infrastructure references. The eventId/occurredAt are stamped at
// construction so events are immutable from creation.
import { randomUUID } from "crypto";

export interface DomainEventMetadata {
  eventId: string;
  occurredAt: string;
  eventName: string;
  aggregateId: string;
  version: number;
}

export abstract class DomainEvent<TPayload = unknown> {
  readonly metadata: DomainEventMetadata;
  readonly payload: TPayload;

  protected constructor(args: {
    eventName: string;
    aggregateId: string;
    payload: TPayload;
    version?: number;
  }) {
    this.metadata = {
      eventId: randomUUID(),
      occurredAt: new Date().toISOString(),
      eventName: args.eventName,
      aggregateId: args.aggregateId,
      version: args.version ?? 1,
    };
    this.payload = args.payload;
  }
}

File: services/ai-agents/app/domain/shared/domain_event.py
from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any


@dataclass(frozen=True, slots=True)
class DomainEventMetadata:
    event_id: str
    occurred_at: str
    event_name: str
    aggregate_id: str
    version: int


@dataclass(frozen=True, slots=True)
class DomainEvent:
    """
    Base for events the domain emits. Concrete events extend this with
    a typed payload; metadata is stamped at construction.
    """

    metadata: DomainEventMetadata
    payload: dict[str, Any] = field(default_factory=dict)

    @staticmethod
    def make_metadata(*, event_name: str, aggregate_id: str, version: int = 1) -> DomainEventMetadata:
        return DomainEventMetadata(
            event_id=str(uuid.uuid4()),
            occurred_at=datetime.now(timezone.utc).isoformat(),
            event_name=event_name,
            aggregate_id=aggregate_id,
            version=version,
        )
